README.md
# IP::Geolocation::MMDB

A Perl module that reads MaxMind DB files and maps IP addresses to location
information such as country and city names.

    use IP::Geolocation::MMDB;
    my $db = IP::Geolocation::MMDB->new(file => 'Country.mmdb');
    my $metadata = $db->metadata;
    my $data = $db->record_for_address('1.2.3.4');
    my $country_code = $db->getcc('2620:fe::9');

## DEPENDENCIES

Requires libmaxminddb 1.2.0 or newer.

Requires Alien::libmaxminddb from CPAN.  On Windows, Alien::MSYS needs to be
installed.  Requires Math::BigInt version 1.999806, which is distributed with
Perl 5.26 and newer.

Requires an IP to country, city or ASN database in the MaxMind DB file format
from [MaxMind](https://www.maxmind.com/) or [DP-IP.com](https://db-ip.com/).

## INSTALLATION

Run the following commands to install the software:

    perl Makefile.PL
    make
    make test
    make install

Type the following command to see the module usage information:

    perldoc IP::Geolocation::MMDB

## LICENSE AND COPYRIGHT

Copyright 2022 Andreas Vögele

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.


Changes
Revision history for IP::Geolocation::MMDB

1.002  2022-01-24 04:51:57 CET

  - Skip the tests if libmaxminddb is too old

1.001  2022-01-23 07:25:04 CET

  - Require libmaxminddb 1.2.0 or newer

1.000  2022-01-22 09:06:57 CET

  - Depend on Math::BigInt 1.999806 and use as_hex() instead of to_hex()
  - Add a safety check

0.009  2022-01-11 08:08:25 CET

  - Depend on Math::BigInt 1.999811

0.008  2022-01-10 15:07:38 CET

  - Implement iterate_search_tree
  - Fix the code for MMDB_UINT128_IS_BYTE_ARRAY

0.007  2022-01-03 13:19:29 CET

  - Implement IP::Geolocation::MMDB::Metadata

0.006  2022-01-02 04:14:40 CET

  - Remove loop initial declarations from MMDB.xs
  - _to_bigint is now a virtual method
  - Improved tests

0.005  2021-12-31 21:16:14 CET

  - Handle MMDB_UINT128_IS_BYTE_ARRAY

0.004  2021-12-31 17:00:26 CET

  - Store 64- and 128-bit integers in Math::BigInt objects

0.003  2021-12-30 10:04:58 CET

  - Don't call gai_strerror(), which may return UTF-16 on Windows
  - Add more tests

0.002  2021-12-29 01:16:55 CET

  - Skip the IPv6 tests on Windows

0.001  2021-12-27 08:03:18 CET

  - First version



Makefile.PL
# This Makefile.PL for IP-Geolocation-MMDB was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.49.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.016;
use ExtUtils::MakeMaker;

die "OS unsupported\n" if $^O eq 'MSWin32';
use Config;
use Alien::libmaxminddb;
die "OS unsupported\n" if Alien::libmaxminddb->version =~ m{^(0|1\.[01])\.};

my %WriteMakefileArgs = (
  "ABSTRACT" => "Read MaxMind DB files",
  "AUTHOR" => "Andreas V\x{f6}gele <voegelas\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "Alien::libmaxminddb" => 0,
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "IP-Geolocation-MMDB",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.016",
  "NAME" => "IP::Geolocation::MMDB",
  "PREREQ_PM" => {
    "Math::BigInt" => "1.999806",
    "XSLoader" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec::Functions" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

%WriteMakefileArgs = (
    %WriteMakefileArgs,
    CCFLAGS => Alien::libmaxminddb->cflags . ' ' . $Config{ccflags},
    LIBS => [ Alien::libmaxminddb->libs ],
);

my %FallbackPrereqs = (
  "File::Spec::Functions" => 0,
  "Math::BigInt" => "1.999806",
  "Test::More" => 0,
  "XSLoader" => 0,
  "utf8" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/IP/Geolocation/MMDB/Metadata.pm
package IP::Geolocation::MMDB::Metadata;

# SPDX-License-Identifier: Artistic-1.0-Perl OR GPL-1.0-or-later

use 5.016;
use warnings;
use utf8;

our $VERSION = 1.002;

sub new {
  my ($class, %attrs) = @_;

  my $self = bless \%attrs, $class;

  return $self;
}

sub binary_format_major_version {
  my ($self) = @_;

  return $self->{binary_format_major_version};
}

sub binary_format_minor_version {
  my ($self) = @_;

  return $self->{binary_format_minor_version};
}

sub build_epoch {
  my ($self) = @_;

  return $self->{build_epoch};
}

sub database_type {
  my ($self) = @_;

  return $self->{database_type};
}

sub languages {
  my ($self) = @_;

  return $self->{languages};
}

sub description {
  my ($self) = @_;

  return $self->{description};
}

sub ip_version {
  my ($self) = @_;

  return $self->{ip_version};
}

sub node_count {
  my ($self) = @_;

  return $self->{node_count};
}

sub record_size {
  my ($self) = @_;

  return $self->{record_size};
}

1;
__END__

=encoding UTF-8

=head1 NAME

IP::Geolocation::MMDB::Metadata - Metadata from a MaxMind DB file

=head1 VERSION

version 1.002

=head1 SYNOPSIS

  use IP::Geolocation::MMDB;
  my $db = IP::Geolocation::MMDB->new(file => 'City.mmdb');
  my $metadata = $db->metadata;

=head1 DESCRIPTION

A class for metadata from a MaxMind DB file.

=head1 SUBROUTINES/METHODS

=head2 new

  my $metadata = IP::Geolocation::MMDB::Metadata->new(
    binary_format_major_version => 2,
    binary_format_minor_version => 0,
    build_epoch   => time,
    database_type => 'City',
    languages     => [qw(en fr pt-BR)],
    description   => {
      en => 'IP to city',
      fr => 'IP vers ville',
    },
    ip_version    => 6,
    node_count    => 3829268,
    record_size   => 28,
  );

Returns a new metadata object.

=head2 binary_format_major_version

  my $major_version = $metadata->binary_format_major_version;

Returns the database format's major version number.

=head2 binary_format_minor_version

  my $minor_version = $metadata->binary_format_minor_version;

Returns the database format's minor version number.

=head2 build_epoch

  my $t = gmtime $metadata->build_epoch;

Returns the database's build timestamp as an epoch number.

=head2 database_type

  my $database_type = $metadata->database_type;

Returns a free-form string indicating the database type.

=head2 languages

  for my $language (@{$metadata->languages}) {
    say $language;
  }

Returns a reference to an array of locale codes indicating what languages this
database has information for.

=head2 description

  my %description_for = %{$metadata->description};
  for my $language (keys %description_for) {
    my $description = $description_for{$language};
    say "$language: $description";
  }

Returns a reference to a hash that maps locale codes to strings that describe
the database content.

=head2 ip_version

  my $ip_version = $metadata->ip_version;

Returns 4 or 6.

=head2 node_count

  my $node_count = $metadata->node_count;

Returns the number of nodes in the database's search tree.

=head2 record_size

  my $record_size = $metadata->record_size;

Returns the record size for nodes in the database's search tree.

=head1 DIAGNOSTICS

None.

=head1 CONFIGURATION AND ENVIRONMENT

None.

=head1 DEPENDENCIES

None.

=head1 INCOMPATIBILITIES

None.

=head1 SEE ALSO

L<IP::Geolocation::MMDB>

=head1 AUTHOR

Andreas Vögele E<lt>voegelas@cpan.orgE<gt>

=head1 BUGS AND LIMITATIONS

None known.

=head1 LICENSE AND COPYRIGHT

Copyright 2022 Andreas Vögele

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/IP/Geolocation/MMDB.pm
package IP::Geolocation::MMDB;

# SPDX-License-Identifier: Artistic-1.0-Perl OR GPL-1.0-or-later

use 5.016;
use warnings;
use utf8;

our $VERSION = 1.002;

use IP::Geolocation::MMDB::Metadata;
use Math::BigInt 1.999806;

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

sub new {
  my ($class, %attrs) = @_;

  my $file  = $attrs{file} or die q{The "file" parameter is mandatory};
  my $flags = 0;

  my $self = $class->_new($file, $flags);

  return $self;
}

sub getcc {
  my ($self, $ip_address) = @_;

  my $country_code;

  my $data = $self->record_for_address($ip_address);
  if (ref $data eq 'HASH') {
    if (exists $data->{country}) {
      my $country = $data->{country};
      if (exists $country->{iso_code}) {
        $country_code = $country->{iso_code};
      }
    }
  }

  return $country_code;
}

sub metadata {
  my ($self) = @_;

  return IP::Geolocation::MMDB::Metadata->new(%{$self->_metadata});
}

## no critic (Subroutines::ProhibitUnusedPrivateSubroutines)

sub _to_bigint {
  my ($self, $bytes) = @_;

  return Math::BigInt->from_bytes($bytes);
}

1;
__END__

=encoding UTF-8

=head1 NAME

IP::Geolocation::MMDB - Read MaxMind DB files

=head1 VERSION

version 1.002

=head1 SYNOPSIS

  use IP::Geolocation::MMDB;
  my $db = IP::Geolocation::MMDB->new(file => 'Country.mmdb');
  my $metadata = $db->metadata;
  my $data = $db->record_for_address('1.2.3.4');
  my $country_code = $db->getcc('2620:fe::9');

=head1 DESCRIPTION

A Perl module that reads MaxMind DB files and maps IP addresses to location
information such as country and city names.

=head1 SUBROUTINES/METHODS

=head2 new

  my $db = IP::Geolocation::MMDB->new(file => 'Country.mmdb');

Returns a new database object.  Dies if the specified file cannot be read.

=head2 getcc

  my $country_code = $db->getcc($ip_address);

Takes an IPv4 or IPv6 address as a string and returns a two-letter country
code or the undefined value.  Dies if the address is not a valid IP address.

=head2 record_for_address

  my $data = $db->record_for_address($ip_address);

Takes an IPv4 or IPv6 address as a string and returns the data associated with
the IP address or the undefined value.  Dies if the address is not a valid IP
address.

The returned data is usually a hash reference but could also be a an array
reference or a scalar for custom databases.  Here's an example from an IP to
city database:

  {
    city => {
      geoname_id => 2950159,
      names      => {
        en => "Berlin"
      }
    },
    country => {
      geoname_id => 2921044,
      iso_code   => "DE",
      names      => {
        en => "Germany",
        fr => "Allemagne"
      }
    },
    location => {
      latitude  => 52.524,
      longitude => 13.411
    }
  }

=head2 iterate_search_tree

  sub data_callback {
    my ($numeric_ip, $prefix_length, $data) = @_;
  }

  sub node_callback {
    my ($node_number, $left_node_number, $right_node_number) = @_;
  }

  $db->iterate_search_tree(\&data_callback, \&node_callback);

Iterates over the entire search tree.  Calls the provided callbacks for each
data record and node in the tree.  Both callbacks are optional.

The data callback is called with a numeric IP address as a L<Math::BigInt>
object, a network prefix length and the data associated with the network.

The node callback is called with the node's number in the tree and the
children's node numbers.

=head2 metadata

  my $metadata = $db->metadata;

Returns an L<IP::Geolocation::MMDB::Metadata> object for the database.

=head2 libmaxminddb_version

  my $version = IP::Geolocation::MMDB::libmaxminddb_version;

Returns the libmaxminddb version.

=head1 DIAGNOSTICS

=over

=item B<< The "file" parameter is mandatory >>

The constructor was called without a database filename.

=item B<< Error opening database file >>

The database file could not be read.

=item B<< The IP address you provided is not a valid IPv4 or IPv6 address >>

A parameter did not contain a valid IP address.

=item B<< Error looking up IP address >>

A database error occurred while looking up an IP address.

=item B<< Entry data error looking up >>

A database error occurred while reading the data associated with an IP
address.

=item B<< Error getting metadata >>

An error occurred while reading the database's metadata.

=item B<< Invalid record when reading node >>

Either an invalid node was looked up or the database is corrupt.

=item B<< Unknown record type >>

An unknown record type was found in the database.

=item B<< Invalid depth when reading node >>

An error occurred while traversing the search tree.

=back

=head1 CONFIGURATION AND ENVIRONMENT

None.

=head1 DEPENDENCIES

Requires L<Alien::libmaxminddb> from CPAN.  Requires L<Math::BigInt> version
1.999806, which is distributed with Perl 5.26 and newer.

Requires libmaxminddb 1.2.0 or newer.

Requires an IP to country, city or ASN database in the MaxMind DB file format
from L<MaxMind|https://www.maxmind.com/> or L<DP-IP.com|https://db-ip.com/>.

Windows is not supported.  Please do not ask for Windows support.

=head1 INCOMPATIBILITIES

None.

=head1 AUTHOR

Andreas Vögele E<lt>voegelas@cpan.orgE<gt>

=head1 BUGS AND LIMITATIONS

If your Perl interpreter does not support 64-bit integers,
MMDB_DATA_TYPE_UINT64 values are put into Math::BigInt objects;

MMDB_DATA_TYPE_UINT128 values are put into Math::BigInt objects;

=head1 LICENSE AND COPYRIGHT

Copyright 2022 Andreas Vögele

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
